File: common/logisticspipes/modules/AsyncQuicksortModule.java
package logisticspipes.modules;

import java.util.List;

import net.minecraft.core.Direction;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;

import net.neoforged.neoforge.network.PacketDistributor;

import org.jspecify.annotations.Nullable;

import logisticspipes.LPConfigs;
import logisticspipes.api.property.Property;
import logisticspipes.interfaces.IInventoryUtil;
import logisticspipes.interfaces.IPipeServiceProvider;
import logisticspipes.logistics.PassiveSinkFinder;
import logisticspipes.network.ModuleTarget;
import logisticspipes.network.to_client.module.QuickSortStateMessage;
import logisticspipes.particle.Particles;
import logisticspipes.pipes.basic.CoreRoutedPipe;
import logisticspipes.routing.AsyncRouting;
import logisticspipes.routing.ServerRouter;
import logisticspipes.utils.ItemUtil;
import logisticspipes.utils.LPExecutors;
import logisticspipes.utils.PlayerCollectionList;
import logisticspipes.utils.SinkReply;
import logisticspipes.utils.item.ItemIdentifier;

public class AsyncQuicksortModule extends AsyncModule<@Nullable AsyncQuicksortModule.QuicksortSetup,
    @Nullable AsyncQuicksortModule.QuicksortResult> {

    public static final int STALLED_DELAY = 24;
    public static final int NORMAL_DELAY = 6;

    private static final String NAME = "quick_sort";

    /** The slot and what sat in it when the job was planned. */
    public record QuicksortSetup(int slot, ItemStack stack) {}

    /** Where the planned stack is headed, as worked out away from the server thread. */
    public record QuicksortResult(int slot, ItemIdentifier itemid, int destRouterId, SinkReply sinkReply) {}

    private final PlayerCollectionList localSlotWatchers = new PlayerCollectionList();
    private boolean stalled = true;
    /**
     * The slot the next tick will look at. Watchers are told which slot is being *worked on*, so
     * the message is sent where the work starts and not from a setter here: an increment goes
     * through the setter too, and would announce the slot after the one in hand -- leaving the
     * first slot of the inventory announced and overwritten within the same tick, so never drawn.
     */
    private int currentSlot = 0;
    private int stallSlot = 0;

    public static String getName() {
        return NAME;
    }

    @Override
    public List<Property<?>> getProperties() {
        return List.of();
    }

    private int getEnergyPerStack() {
        return 500 + 1000 * getUpgradeManager().getItemStackExtractionUpgrade();
    }

    @Override
    public int getEveryNthTick() {
        return stalled ? STALLED_DELAY : NORMAL_DELAY;
    }

    @Override
    public String getLPName() {
        return NAME;
    }

    @Override
    public @Nullable QuicksortSetup jobSetup() {
        IPipeServiceProvider service = this.service;
        if (service == null || !(service.getRouter() instanceof ServerRouter serverRouter)) {
            return null;
        }
        List<@Nullable IInventoryUtil> inventories = PipeServiceProviderUtil.availableInventories(service);
        IInventoryUtil inventory = inventories.isEmpty() ? null : inventories.getFirst();
        if (inventory == null || inventory.getContainerSize() == 0) {
            return null;
        }
        if (currentSlot >= inventory.getContainerSize()) {
            currentSlot = 0;
        }
        int slot = currentSlot;
        currentSlot = slot + 1;
        notifyWorkingSlot(slot);
        ItemStack stack = inventory.getItem(slot);
        if (!stalled && slot == stallSlot) {
            stalled = true;
        }
        if (stack.isEmpty()) {
            return null;
        }
        AsyncRouting.updateServerRouterLsa(serverRouter);
        if (!LPConfigs.COMMON.DISABLE_ASYNC_WORK.getAsBoolean()
            && AsyncRouting.needsRoutingTableUpdate(serverRouter)) {
            // go async
            return new QuicksortSetup(slot, stack);
        }
        ItemIdentifier itemid = ItemIdentifier.get(stack);
        PassiveSinkFinder.Destination destination =
            PassiveSinkFinder.getDestination(stack, itemid, false, serverRouter, List.of());
        if (destination == null) {
            return null;
        }
        extractAndSend(slot, stack, inventory, destination.routerId(), destination.sinkReply());
        return null;
    }

    @Override
    public @Nullable QuicksortResult tickAsync(@Nullable QuicksortSetup setupObject) {
        if (setupObject == null) {
            return null;
        }
        IPipeServiceProvider service = this.service;
        if (service == null || !(service.getRouter() instanceof ServerRouter serverRouter)) {
            return null;
        }
        AsyncRouting.updateRoutingTable(serverRouter);
        ItemIdentifier itemid = ItemIdentifier.get(setupObject.stack());
        // Touches routers and pipes, which only the server thread may do.
        PassiveSinkFinder.Destination destination = LPExecutors.onServerThread(() ->
            PassiveSinkFinder.getDestination(setupObject.stack(), itemid, false, serverRouter, List.of()));
        if (destination == null) {
            return null;
        }
        return new QuicksortResult(setupObject.slot(), itemid, destination.routerId(), destination.sinkReply());
    }

    @Override
    public void completeJob(@Nullable QuicksortResult result) {
        IPipeServiceProvider service = this.service;
        if (result == null || service == null) {
            return;
        }
        List<@Nullable IInventoryUtil> inventories = PipeServiceProviderUtil.availableInventories(service);
        IInventoryUtil inventory = inventories.isEmpty() ? null : inventories.getFirst();
        if (inventory == null || result.slot() >= inventory.getContainerSize()) {
            return;
        }
        ItemStack stack = inventory.getItem(result.slot());
        if (ItemUtil.equalsWithNBT(result.itemid(), stack)) {
            extractAndSend(result.slot(), stack, inventory, result.destRouterId(), result.sinkReply());
        }
    }

    private void extractAndSend(int slot, ItemStack stack, IInventoryUtil inventory, int destRouterId,
        SinkReply sinkReply) {
        IPipeServiceProvider service = this.service;
        if (service == null) {
            return;
        }
        Direction pointedOrientation = service.getPointedOrientation();
        if (pointedOrientation == null) {
            return;
        }
        int toExtract = ItemUtil.getExtractionMax(stack.getCount(), stack.getMaxStackSize(), sinkReply);
        if (toExtract <= 0 || !service.useEnergy(getEnergyPerStack())) {
            return;
        }
        stalled = false;
        stallSlot = slot;
        ItemStack extracted = inventory.removeItem(slot, toExtract);
        if (extracted.isEmpty()) {
            return;
        }
        service.sendStack(extracted, destRouterId, sinkReply, CoreRoutedPipe.ItemSendMode.Fast, pointedOrientation);
        service.spawnParticle(Particles.ORANGE_SPARKLE, 8);
    }

    @Override
    public void runSyncWork() {}

    @Override
    public boolean receivePassive() {
        return false;
    }

    @Override
    public boolean hasGenericInterests() {
        return false;
    }

    @Override
    public boolean interestedInUndamagedID() {
        return false;
    }

    @Override
    public boolean interestedInAttachedInventory() {
        return false;
    }

    private void notifyWorkingSlot(int slot) {
        localSlotWatchers.send(new QuickSortStateMessage(ModuleTarget.of(this), slot));
    }

    public void addWatchingPlayer(Player player) {
        localSlotWatchers.add(player);
        if (player instanceof ServerPlayer serverPlayer) {
            // Where the last tick was, which is the slot that is currently ringed for everyone
            // else watching.
            PacketDistributor.sendToPlayer(serverPlayer,
                new QuickSortStateMessage(ModuleTarget.of(this), Math.max(currentSlot - 1, 0)));
        }
    }

    public void removeWatchingPlayer(Player player) {
        localSlotWatchers.remove(player);
    }
}

File: src/main/kotlin/network/rs485/logisticspipes/module/AsyncQuicksortModule.kt
/*
 * Copyright (c) 2020  RS485
 *
 * "LogisticsPipes" is distributed under the terms of the Minecraft Mod Public
 * License 1.0.1, or MMPL. Please check the contents of the license located in
 * https://github.com/RS485/LogisticsPipes/blob/dev/LICENSE.md
 *
 * This file can instead be distributed under the license terms of the
 * MIT license:
 *
 * Copyright (c) 2020  RS485
 *
 * This MIT license was reworded to only match this file. If you use the regular
 * MIT license in your project, replace this copyright notice (this line and any
 * lines below and NOT the copyright line above) with the lines from the original
 * MIT license located here: http://opensource.org/licenses/MIT
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this file and associated documentation files (the "Source Code"), to deal in
 * the Source Code without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
 * of the Source Code, and to permit persons to whom the Source Code is furnished
 * to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Source Code, which also can be
 * distributed under the MIT.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package network.rs485.logisticspipes.module

import network.rs485.grow.Coroutines
import network.rs485.logisticspipes.logistics.LogisticsManager
import network.rs485.logisticspipes.property.Property
import network.rs485.logisticspipes.util.equalsWithNBT
import network.rs485.logisticspipes.util.getExtractionMax
import logisticspipes.LPConfigs
import logisticspipes.interfaces.IInventoryUtil
import logisticspipes.modules.PipeServiceProviderUtil
import logisticspipes.network.ModuleTarget
import logisticspipes.network.to_client.module.QuickSortStateMessage
import logisticspipes.particle.Particles
import logisticspipes.pipes.basic.CoreRoutedPipe
import logisticspipes.routing.AsyncRouting
import logisticspipes.routing.ServerRouter
import logisticspipes.utils.PlayerCollectionList
import logisticspipes.utils.SinkReply
import logisticspipes.utils.item.ItemIdentifier
import net.neoforged.neoforge.network.PacketDistributor
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.withContext

const val STALLED_DELAY = 24
const val NORMAL_DELAY = 6

data class QuicksortAsyncResult(
    val slot: Int,
    val itemid: ItemIdentifier,
    val destRouterId: Int,
    val sinkReply: SinkReply,
)

class AsyncQuicksortModule : AsyncModule<Pair<Int, ItemStack>?, QuicksortAsyncResult?>() {
    companion object {
        @JvmStatic
        val name: String = "quick_sort"
    }

    private val localSlotWatchers = PlayerCollectionList()
    private var stalled = true
    private var currentSlot = 0
        set(value) {
            field = value
            localSlotWatchers.send(QuickSortStateMessage(ModuleTarget.of(this), value))
        }
    private var stallSlot = 0

    override val properties: List<Property<*>>
        get() = emptyList()

    private val energyPerStack: Int
        get() = upgradeManager.let { 500 + 1000 * it.itemStackExtractionUpgrade }.toInt()
    override val everyNthTick: Int
        get() = if (stalled) STALLED_DELAY else NORMAL_DELAY

    override fun getLPName(): String = name

    override fun jobSetup(): Pair<Int, ItemStack>? {
        val serverRouter = this.service?.router as? ServerRouter ?: return null
        val inventory = service?.let { PipeServiceProviderUtil.availableInventories(it) }?.firstOrNull() ?: return null
        if (inventory.containerSize == 0) return null
        if (currentSlot >= inventory.containerSize) currentSlot = 0
        val slot = currentSlot++
        val stack = inventory.getItem(slot)
        if (!stalled && slot == stallSlot) stalled = true
        if (stack.isEmpty) return null
        AsyncRouting.updateServerRouterLsa(serverRouter)
        if (!LPConfigs.COMMON.DISABLE_ASYNC_WORK.asBoolean && AsyncRouting.needsRoutingTableUpdate(serverRouter)) {
            // go async
            return slot to stack
        }
        val itemid = ItemIdentifier.get(stack)
        val result = LogisticsManager.getDestination(stack, itemid, false, serverRouter, emptyList()) ?: return null
        extractAndSend(slot, stack, inventory, result.first, result.second)
        return null
    }

    override suspend fun tickAsync(setupObject: Pair<Int, ItemStack>?): QuicksortAsyncResult? {
        if (setupObject == null) return null
        val serverRouter = this.service?.router as? ServerRouter ?: return null
        AsyncRouting.updateRoutingTable(serverRouter)
        val itemid = ItemIdentifier.get(setupObject.second)
        val result = withContext(Coroutines.serverScope.coroutineContext) {
            LogisticsManager.getDestination(setupObject.second, itemid, false, serverRouter, emptyList())
        } ?: return null
        return QuicksortAsyncResult(setupObject.first, itemid, result.first, result.second)
    }

    @ExperimentalCoroutinesApi
    override fun completeJob(deferred: Deferred<QuicksortAsyncResult?>) {
        val result = deferred.getCompleted() ?: return
        val inventory = service?.let { PipeServiceProviderUtil.availableInventories(it) }?.firstOrNull() ?: return
        if (result.slot >= inventory.containerSize) return
        val stack = inventory.getItem(result.slot)
        if (result.itemid.equalsWithNBT(stack)) {
            extractAndSend(result.slot, stack, inventory, result.destRouterId, result.sinkReply)
        }
    }

    private fun extractAndSend(
        slot: Int,
        stack: ItemStack,
        inventory: IInventoryUtil,
        destRouterId: Int,
        sinkReply: SinkReply,
    ) {
        val service = service ?: return
        val pointedOrientation = service.pointedOrientation ?: return
        val toExtract = getExtractionMax(stack.count, stack.maxStackSize, sinkReply)
        if (toExtract <= 0) return
        if (!service.useEnergy(energyPerStack)) return
        stalled = false
        stallSlot = slot
        val extracted = inventory.removeItem(slot, toExtract)
        if (extracted.isEmpty) return
        service.sendStack(
            extracted,
            destRouterId,
            sinkReply,
            CoreRoutedPipe.ItemSendMode.Fast,
            pointedOrientation,
        )
        service.spawnParticle(Particles.ORANGE_SPARKLE, 8)
    }

    override fun runSyncWork() {}

    override fun receivePassive(): Boolean = false

    override fun hasGenericInterests(): Boolean = false

    override fun interestedInUndamagedID(): Boolean = false

    override fun interestedInAttachedInventory(): Boolean = false

    fun addWatchingPlayer(player: Player) {
        localSlotWatchers.add(player)
        if (player is ServerPlayer) {
            PacketDistributor.sendToPlayer(player, QuickSortStateMessage(ModuleTarget.of(this), currentSlot))
        }
    }

    fun removeWatchingPlayer(player: Player) {
        localSlotWatchers.remove(player)
    }

}
